### PR TITLE
Time stepper schedule

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
## Proposed changes

This PR goes after #1732. The new class `TimeScheduler` is introduced to replace `TimeManager`.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/time_stepper/scheduler.py [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+            atol=atol,
+        )
+
+
+def _log_schedule_interval_start(t_start: float, t_end: float, name: str = "") -> None:
+    log_message = "Reached new schedule interval"
+    if name != "":
+        log_message += f' "{name}"'
+    delta = t_end - t_start
+    log_message += f" [{t_start:.1e}, {t_start:.1e} + {delta:.1e})."
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/scheduler.py]: Love the formatting! Consider  [{t_start:.1e}, {t_start:.1e} + {delta:.1e} = {t_start + delta:.1e}). to get best of both?

Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/scheduler.py]: Whatever floats your boat!

File: src/porepy/time_stepper/time_step_constraint.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            if num_iter <= self.iter_min:
+                return dt * self.increase_factor
+            elif num_iter >= self.iter_max:
+                # Decrease dt, but not below dt_min.
+                return max(dt * self.decrease_factor, self.dt_min)
+            else:
+                return dt
+        else:
+            if abs(dt - self.dt_min) < self.t_snap:
+                raise CannotRecomputeTimeStep(
```
Comment (Yuriyzabegaev) [inline src/porepy/time_stepper/time_step_constraint.py]: This bothers me and we should discuss it in person.

File: src/porepy/time_stepper/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+        # t_end.
+        next_checkpoint = (
+            self.schedule.t_end if next_interval is None else next_interval.t_start
+        )
+
+        # Dt should be not larger than the interval's dt_max.
+        dt = min(dt, current_interval.dt_max)
+        # If constraints made dt smaller than the interval's dt_min, abort simulation.
+        if dt < current_interval.dt_min:
+            raise CannotRecomputeTimeStep(
```
Comment (Yuriyzabegaev) [inline src/porepy/time_stepper/scheduler.py:240]: I'm not fully comfortable with this, because it means if CourantConstraint forces dt less than dt_min (due to very high velosity), it will also abort simulation. But maybe it's not a bad thing?

File: src/porepy/time_stepper/time_step_control.py [hunk 4]
```diff
@@ -1,400 +1,198 @@
...
         constant_dt: bool = False,
-        dt_min_max: Optional[tuple[Union[int, float], Union[int, float]]] = None,
-        iter_max: int = 15,
+        dt_min_max: Optional[tuple[pp.number, pp.number]] = None,
+        iter_max: Optional[int] = None,
         iter_optimal_range: tuple[int, int] = (4, 7),
         iter_relax_factors: tuple[float, float] = (0.7, 1.3),
         recomp_factor: float = 0.5,
-        recomp_max: int = 10,
-        print_info: bool = False,
```
Comment (Yuriyzabegaev) [inline src/porepy/time_stepper/time_step_control.py:211]: I deleted parameter print_info, because it was ignored anyway.

File: src/porepy/time_stepper/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+
+
+class TimeScheduler(TimeSchedulerBase):
+    """Non-constant time step scheduler.
+
+    Each schedule interval may contain its own set of time step constraints, that
+    adjust dt based on simulation context. Each constraint suggests the new dt value,
+    which may be smaller or larger than the current dt. The minimum of the suggestions
+    is applied. If the minimum is below the interval's `dt_min`, the simulation is
+    stopped.
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/scheduler.py]: If the suggestion is above max, we cap, right? Document this?

File: src/porepy/time_stepper/scheduler.py (hunk 3 above)
Comment (IvarStefansson) [reply src/porepy/time_stepper/scheduler.py:239]: At the moment of writing, that seems sensible to me. That means that if the user expects (or can't rule out) high velocities, they must allow a small dt min.

File: src/porepy/time_stepper/scheduler.py [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    if time_manager.dt_min_max is None:
+        dt_min = dt_max = None
+    else:
+        dt_min, dt_max = time_manager.dt_min_max
+    if dt_min is None:
+        dt_min = float(dt_init) * 1e-3
+    if dt_max is None:
+        dt_max = float(dt_init) * 1e3
+
+    if not time_manager.is_constant:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/scheduler.py:262]: How would I avoid having these if I don't want any TargetNonlinearIterations at all?

File: src/porepy/time_stepper/scheduler.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+
+    Parameters:
+        intervals: List of intervals. The last interval is `[t_last, ∞)`.
+        atol: Snapping time. Time differences below it are treated as zero.
+
+    """
+
+    def __init__(self, intervals: list[TimeInterval], atol: float) -> None:
+        self.intervals = intervals
+        self._interval_starts = [interval.t_start - atol for interval in intervals]
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/scheduler.py:296]: This is a bit counterintuitive. Please add a comment.

File: src/porepy/time_stepper/time_step_status.py [hunk 8]
```diff
@@ -70,26 +78,53 @@ def serialize(self) -> str:
...
+    """
 
     def serialize(self) -> str:
         return "successful"
 
+    @property
+    def dt(self) -> float:
+        """Simulation time step magnitude."""
+        if len(self.attempts) == 0:
+            raise ValueError
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_status.py]: Would returning 0 be dangerous?

File: src/porepy/viz/solver_statistics.py [hunk 9]
```diff
@@ -385,14 +379,17 @@ def append_global_data(self, data: dict[str, dict]) -> dict[str, dict]:
         # Extract final convergence status.
         final_convergence_status = _leafs_only(self.convergence_status.to_str())
 
-        # Determine number of waisted iterations.
+        # Determine number of wasted iterations.
         # TODO: Rethink during upgrade of time integration.
         total_num_waisted_iterations = 0
-        for simulation_status, num_iterations in zip(
+        for simulation_status, num_iterations in zip(  # TODO YZ: Remove num iterations.
```
Comment (IvarStefansson) [inline src/porepy/viz/solver_statistics.py]: Should this be removed in this PR?

File: tests/models/test_model_runner.py [hunk 10]
```diff
@@ -79,3 +79,42 @@ def solve_linear_system(
...
+    already during prepare_simulation(), not just after the first time step.
+
+    ModelRunner.__init__ must resolve the passed-in time_stepper and seed
+    model.time_data from its scheduler *before* calling prepare_simulation(). Otherwise,
+    anything invoked during prepare_simulation() that depends on self.time_data.schedule
+    (e.g. time-dependent boundary conditions defined per schedule point) would
+    incorrectly see the SolutionStrategy.__init__ placeholder schedule [0.0, 1.0]
+    instead of the real one.
+
+    TODO: Written based on discovery pattern. Consider less specific test.
```
Comment (IvarStefansson) [inline tests/models/test_model_runner.py]: Could we not do something like overwriting prepare_simulation to check the time stepper and time data?

File: tests/viz/test_data_saving_model_mixin.py [hunk 11]
```diff
@@ -25,9 +26,24 @@
 from porepy.viz.data_saving_model_mixin import FractureDeformationExporting
 
 
+@pytest.fixture
+def data_folder(request):
```
Comment (IvarStefansson) [inline tests/viz/test_data_saving_model_mixin.py:30]: Thanks, I guess. But might this have been fixed in a separate upstream PR?

File: src/porepy/time_stepper/time_step_constraint.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    @abstractmethod
+    def suggest_dt(self, dt: float, context: dict) -> float:
+        """The constraint suggests the maximum dt, permitted for the next time step.
+
+        In other words, for the returned `dt_new` the permitted time step range is
+        `(0, dt_new]`.
+
+        Parameters:
+            dt: The time step magnitude that was used for the current time step attempt.
+            context: Simulation context to make the adjustment based on. Implementations
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py:34]: It seems to be a bit cumbersome to change what is passed down in context. Is this a problem to be tackled if and when we need more context?

File: src/porepy/time_stepper/time_step_constraint.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    Expects "nonlinear_solver_status" (:class:`pp.solvers.NonlinearSolverStatus`),
+    "t_snap" and "dt_min" in context.
+
+    If dt is to be decreased due to failure and it is about to become smaller than
+    `dt_min`, it will first decrease it to `dt_min`. If the attempt with dt = dt_min
+    fails, it will decrease dt below `dt_min`, which will force the TimeScheduler to
+    stop the simulation.
+
+    Parameters:
+        iter_min: Left bound of the nonlinear solver's desired iterations range.
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py:63]: Do we need to define here and when setting attributes? I thought the latter sufficed.

File: src/porepy/time_stepper/time_step_constraint.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        """Velocity tolerance, below treated as zero."""
+
+    def suggest_dt(self, dt: float, context: dict) -> float:
+        model = context.get("model", None)
+        if not isinstance(model, pp.SolutionStrategy):
+            raise ValueError("CourantTimeStepConstraint requires 'model' in context.")
+
+        dt = float("inf")
+        for subdomain in model.mdg.subdomains():
+            # TODO: Mobility is not included, unit of darcy_flux is [m^2 / s].
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py]: Please resolve todo

File: src/porepy/time_stepper/time_step_control.py [hunk 15]
```diff
@@ -1,400 +1,198 @@
...
+            :class:`pp.time_stepper.TimeScheduler`.
+        recomp_factor: Factor of how to decrease dt if the nonlinear solver failed and
+            the time step is to be recomputed. Passed to
+            :class:`pp.time_stepper.TimeScheduler`.
+        recomp_max: Deprecated, does nothing. Control the number of attempts to make a
+            time step in :class:`pp.time_stepper.TimeStepper`.
+        rtol: Deprecated, does nothing.
+        atol: Snapping time. If the time difference is below it, treats two time points
+            as equal.
+        advanced_schedule: A more advanced way to define the simulation schedule. See
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_control.py]: This one confuses me. Please check if it is properly documented and point me to the right place.

File: src/porepy/time_stepper/time_stepper.py [hunk 16]
```diff
@@ -80,168 +100,135 @@ def perform_time_step(
...
-        # Cache previous time for trial.
-        previous_time = self.time_manager.time
+        time_manager = model.time_manager
+        accepted_time = time_manager.time
+        accepted_index = time_manager.time_index
+        attempts_data: list[TimeStepperAttemptData] = []
 
-        # Logging time step start.
-        _log_time_step(time_manager=self.time_manager)
+        def rollback_time() -> None:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_stepper.py]: I read this as a verb, which I think implies "roll_back_time". Ignore if I've misunderstood.

File: src/porepy/time_stepper/time_stepper.py [hunk 17]
```diff
@@ -80,168 +100,135 @@ def perform_time_step(
...
-            num_iterations = nonlinear_solver_status.number_of_iterations()
-
-            current_dt = self.time_manager.dt
-            new_time = self.time_manager.time
-            self.time_manager.compute_time_step(iterations=num_iterations)
-            return TimeStepperStatusSuccess(
-                dt=current_dt,
-                time=new_time,
-                nonlinear_solver_status=nonlinear_solver_status,
+            if attempt > 0:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_stepper.py]: Why conditional?

File: src/porepy/time_stepper/time_stepper.py [hunk 18]
```diff
@@ -80,168 +100,135 @@ def perform_time_step(
...
+                    context={
+                        "model": model,
+                        "nonlinear_solver_status": nonlinear_status,
+                    },
                 )
-            except ValueError as e:
-                # Time manager raises a value error if dt cannot be reduced any further.
-                return TimeStepperStatusFailure(
-                    nonlinear_solver_status=nonlinear_solver_status, reason=str(e)
+            except CannotRecomputeTimeStep as exc:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_stepper.py:148]: I love the class name!

File: src/porepy/time_stepper/time_stepper.py [hunk 19]
```diff
@@ -80,168 +100,135 @@ def perform_time_step(
...
+                # Necessary after a successful trial, harmless after done twice.
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_stepper.py]: What does this comment mean? And should we roll back or not in the except case?

File: src/porepy/time_stepper/time_stepper.py [hunk 20]
```diff
@@ -80,168 +100,135 @@ def perform_time_step(
...
+                    model,
+                    TimeStepperStatusSuccess(
+                        time=accepted_time,
+                        attempts=list(attempts_data),
+                    ),
+                )
+            else:
+                _ = _log_and_return_time_step_data(
+                    model,
+                    TimeStepperStatusContinueIterating(
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_stepper.py:179]: Explain why this is continue iterating.

File: tests/time_stepper/test_scheduler.py
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py:1]: Add file docstring summary?

File: tests/time_stepper/test_scheduler.py [hunk 21]
```diff
@@ -0,0 +1,512 @@
...
+        dt_min_max=dt_min_max,
+        iter_optimal_range=nonlinear_iter_optimal_range,
+        iter_relax_factors=nonlinear_iter_relax_factors,
+        recomp_factor=nonlinear_iter_retry_factor,
+        atol=atol,
+    )
+    return assemble_default_time_scheduler(time_manager), time_manager
+
+
+def run_scheduler_collect_data(
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py:57]: This is somewhat sparsely documented. Applies in some other places as well.

File: tests/time_stepper/test_scheduler.py [hunk 22]
```diff
@@ -0,0 +1,512 @@
...
+def test_scheduler_floating_point_inaccuracy(
+    constant_dt: bool, dt: float, dt_snap: float
+):
+    """Test the accumulaton of floating point error when hitting schedule points.
+
+    Purposefully taking weird dt to foster error accumulation. With the default dt_snap
+    (1e-8), the schedule points are registered correctly and well within the margin of
+    error. Problems would start in this example if dt_snap < 1e-15.
+
+    The dt_snap = 1e-50 case tests the behavior when the schedule points stop
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]: In which sense?

File: tests/time_stepper/test_scheduler.py [hunk 23]
```diff
@@ -0,0 +1,512 @@
...
+    registering correctly. Its goal is to ensure that, while they are not registering,
+    the simulation does not abort and continue until completion.
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]:
```suggestion
    the simulation does not abort and continues until completion.
```

File: tests/time_stepper/test_scheduler.py [hunk 24]
```diff
@@ -0,0 +1,512 @@
...
+    [
+        2.5 - 1e-8,
+        2.5 + 1e-8,
+        2.5e8 - 1,
+        2.5e8 + 1,
+        2.5e-8 + 1e-16,
+        2.5e-8 - 1e-16,
+    ],
+)
+def test_scheduler_floating_point_inaccuracy(
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]: I find this test a bit difficult to intuit. Would it help to break it up (possibly extracting helpers), so that each subset of cases could have a more detailed and unified explanation of what is being tested?

File: tests/time_stepper/test_scheduler.py [hunk 25]
```diff
@@ -0,0 +1,512 @@
...
+    else:
+        # General case: All schedule points are handled correctly.
+        np.testing.assert_allclose(
+            checkpoints_hit, expected_schedule, atol=dt_snap, rtol=0
+        )
+
+
+@pytest.mark.parametrize("constant_dt", [True, False])
+def test_inconsistent_schedule(constant_dt: bool):
+    """Inconsistent schedule with constant time step.
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]:
```suggestion
    """Inconsistent schedule.
```

Comment (IvarStefansson) [reply tests/time_stepper/test_scheduler.py]: This reads like two different tests to me.

File: tests/time_stepper/test_scheduler.py [hunk 26]
```diff
@@ -0,0 +1,512 @@
...
+                ],
+                t_end=1.0,
+            ),
+        )
+
+
+@pytest.mark.parametrize("schedule", [[0, 1, 2], [0, 2, 3], [0, 1]])
+def test_initial_time_step_overshoots_schedule_point(schedule: list[int]):
+    with pytest.raises(ValueError):
+        _ = make_default_scheduler(schedule=schedule, dt_init=2.0, constant_dt=True)
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py:372]: The second parametrisation doesn't really overshoot. At least not for non-constant.

File: tests/time_stepper/test_scheduler.py [hunk 27]
```diff
@@ -0,0 +1,512 @@
...
+            "expected_dt": "unreachable",
+            "dt_min": 0.5,
+            "success": False,
+            "should_raise": True,
+        },
+    ],
+)
+def test_target_nonlinear_iterations(case: dict):
+    """Test behaviour of the algorithm when the solution should be recomputed. Note
+    that this should be independent of the number of iterations that the user passes
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]: Does the note only apply to succes=False?

File: tests/time_stepper/test_scheduler.py [hunk 28]
```diff
@@ -0,0 +1,512 @@
...
+        ([0, 10, 20, 30], 1),
+        ([10, 11, 15, 16, 19, 20], 1),
+        (
+            [0, 0.01, 1 * pp.HOUR, 2 * pp.HOUR, 100 * pp.HOUR, 101 * pp.HOUR],
+            2 * pp.HOUR,
+        ),
+    ],
+)
+def test_hitting_schedule_times(schedule, dt_init):
+    """Test if algorithm respects the passed target times from the schedule,"""
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]: Finish docstring.

File: tests/time_stepper/test_scheduler.py [hunk 29]
```diff
@@ -0,0 +1,512 @@
...
+    )
+
+    _, checkpoint_hits = run_scheduler_collect_data(scheduler, time_manager)
+    np.testing.assert_allclose(checkpoint_hits, schedule, atol=t_snap, rtol=0)
+    assert time_manager.final_time_reached()
+
+
+@pytest.mark.parametrize("constant_dt", [True, False])
+def test_time_step_match_schedule_exactly(constant_dt: bool):
+    """Checks the edge case when the dynamic time stepping is on, but the next time
```
Comment (IvarStefansson) [inline tests/time_stepper/test_scheduler.py]: You say dynamic time stepping is on, yet parametrise with constant_dt also True.

File: tests/time_stepper/test_time_step_constraints.py
Comment (IvarStefansson) [inline tests/time_stepper/test_time_step_constraints.py:1]: Why is TargetNonlinearIterations not tested in this file?

**Review by IvarStefansson — COMMENTED** [review]
Very good. Do we need a tutorial/update of existing? And remember to add to changelog.md. I guess you can resolve most comments fairly easily. Let me know if and when we need to discuss some of them.

File: src/porepy/time_stepper/scheduler.py (hunk 5 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/scheduler.py]: documented, thanks

File: src/porepy/time_stepper/scheduler.py (hunk 6 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/scheduler.py:262]: The only way is to assemble it manually and not use `assemble_default_time_scheduler`. But I don't see a case when you ever don't want to have it.

File: src/porepy/time_stepper/scheduler.py (hunk 7 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/scheduler.py:296]: Documented, thanks!

File: src/porepy/time_stepper/time_step_status.py (hunk 8 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_status.py]: Replaced by returning 0. I think it's fine.

File: src/porepy/viz/solver_statistics.py (hunk 9 above)
Comment (Yuriyzabegaev) [reply src/porepy/viz/solver_statistics.py]: missed, thanks.

File: tests/models/test_model_runner.py (hunk 10 above)
Comment (Yuriyzabegaev) [reply tests/models/test_model_runner.py]: There is no time_data anymore. TimeManager (model's property) is the only source of truth. There is no conflict here anymore. However, I decided to keep the test since it's useful and it's not covered anywhere else.

File: tests/viz/test_data_saving_model_mixin.py (hunk 11 above)
Comment (Yuriyzabegaev) [reply tests/viz/test_data_saving_model_mixin.py:30]: Should I make an issue?

File: src/porepy/time_stepper/time_step_constraint.py (hunk 12 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py:34]: Passing additional context will be a problem, yes. It will require implementing your own TimeStepper or subclassing it, or modifying the code in the current TimeStepper. However, the model is passed in the context, so anything that can be fetched from the model is already in the context.

What is not in the context is some custom information about a custom (non-existing yet) nonlinear solver and time stepper. So if someone ever needs to implement them, they will naturally add it to their contexts, without changing the function signature, that's my hope.

File: src/porepy/time_stepper/time_step_constraint.py (hunk 13 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py:63]: I didn't understand the question. This gives you possibility to tailor optimal iteration ranges for different time intervals.

File: src/porepy/time_stepper/time_step_constraint.py (hunk 14 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py]: I'm afraid I don't know how and we need to discuss it

File: src/porepy/time_stepper/time_step_control.py (hunk 15 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_control.py]: We need to discuss it

File: src/porepy/time_stepper/time_stepper.py (hunk 16 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py]: fixed, thanks

File: src/porepy/time_stepper/time_stepper.py (hunk 17 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py]: I thought it's nice to not get spam logs "attempt 1 / 10" (first attempt) for a continuous series of successful attempts. This is subjective, so I'll fix if you like.

File: src/porepy/time_stepper/time_stepper.py (hunk 19 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py]: Replaced the comment with:
```
# If success == False, roll_back_time will be called twice. First after
# failed "solver.solve", second if we fail to recompute the time step
# (here). If success == True, roll_back_time will be called once, here.
# Calling it twice is harmless.
```
Hope it's clear now! I really tried to make this loop easier, but the requirement to call statistics-related routines in the specific points of time, and the fact that the time_scheduler both (i) adjusts dt and (ii) decides whether the simulation should stop really makes it difficult to make this loop more readible.

File: src/porepy/time_stepper/time_stepper.py (hunk 20 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py:179]: Explained, thanks

File: tests/time_stepper/test_scheduler.py
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py:1]: added

File: tests/time_stepper/test_scheduler.py (hunk 22 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: replaced the line with `Problems would start in this example if dt_snap < 1e-15 (schedule points will not register correctly due to error accumulation).`

File: tests/time_stepper/test_scheduler.py (hunk 21 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py:57]: added documentation

File: tests/time_stepper/test_scheduler.py (hunk 23 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: applied in my commit. Thanks

File: tests/time_stepper/test_scheduler.py (hunk 24 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: We should discuss it.. I really don't know how to break it up

File: tests/time_stepper/test_scheduler.py (hunk 25 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: You're right. Split the test into 2.

File: tests/time_stepper/test_scheduler.py (hunk 26 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py:372]: It could overshoot if it didn't decrease the time step to respect the schedule. But it does respect the schedule and does not overshoot. That's what is being tested

File: tests/time_stepper/test_scheduler.py (hunk 27 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: Fixed the docstring, thanks.

File: tests/time_stepper/test_scheduler.py (hunk 28 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: Thanks!

File: tests/time_stepper/test_scheduler.py (hunk 29 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: Thanks! Fixed

File: tests/time_stepper/test_time_step_constraints.py
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_time_step_constraints.py:1]: Because it's tested a lot in test_scheduler as a part of the scheduler. I will move the corresponding test here.

File: src/porepy/time_stepper/time_step_control.py (hunk 15 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_control.py]: See test_time_stepper::test_advanced_scheduler for how to use it

File: src/porepy/time_stepper/time_step_constraint.py (hunk 13 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py:63]: So the question is, whether we should duplicate docstrings for `__init__` parameters and for the class properties?

File: src/porepy/time_stepper/time_stepper.py (hunk 17 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py]: Eirik, we need your vote on this

File: src/porepy/time_stepper/time_step_constraint.py (hunk 14 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py]: From our discussion with Ivar: There are options on what to do with this.

1. Fix it somehow and add a note in the docstring.
2. Rename to "DarcyFluxConstraint" to avoid giving false promises and associations with the CFL.
3. Delete it. In the horizon of 3 months, we don't see if anyone will need it, and if so, it is easily reproduced.

The argument for why not removing it is that it is the 2nd implementation of the TimeStepConstraint abstract class. If doing such abstraction and providing a single implementation, you probably don't need this abstraction. We however know that we want something like this and something similar (if not this) will be useful at some point.

File: tests/time_stepper/test_scheduler.py (hunk 24 above)
Comment (Yuriyzabegaev) [reply tests/time_stepper/test_scheduler.py]: Broke it up after discussion.

File: src/porepy/time_stepper/scheduler.py (hunk 3 above)
Comment (keileg) [reply src/porepy/time_stepper/scheduler.py:239]: I agree, this sounds exactly what dt_min should do.

File: src/porepy/time_stepper/time_step_constraint.py (hunk 13 above)
Comment (keileg) [reply src/porepy/time_stepper/time_step_constraint.py:63]: The instructions in how-to-docstring say yes: https://pmgbergen.github.io/porepy/html/_modules/example_docstrings.html#ExampleClass

File: src/porepy/time_stepper/time_step_constraint.py (hunk 14 above)
Comment (keileg) [reply src/porepy/time_stepper/time_step_constraint.py]: Discuss in person if possible.

File: src/porepy/time_stepper/time_step_control.py [hunk 30]
```diff
@@ -1,400 +1,198 @@
...
+            :class:`pp.time_stepper.TimeScheduler` for details. If None (default), the
+            old way of defining the schedule with array is used. Otherwise, it is
```
Comment (keileg) [inline src/porepy/time_stepper/time_step_control.py]: 'old' is not telling me much here.

File: src/porepy/time_stepper/time_step_control.py (hunk 15 above)
Comment (keileg) [reply src/porepy/time_stepper/time_step_control.py]: At the face of it, I am concerned about the two schedulers. I understand I don't see the full picture, but I would appreciate a discussion here.

File: src/porepy/time_stepper/time_stepper.py (hunk 17 above)
Comment (keileg) [reply src/porepy/time_stepper/time_stepper.py]: This will be at the same line? I vote, weakly, for dropping the conditional and rather having a uniform format of the output.

File: src/porepy/time_stepper/time_step_control.py (hunk 4 above)
Comment (keileg) [reply src/porepy/time_stepper/time_step_control.py:211]: You have my blessing.

File: src/porepy/time_stepper/time_step_constraint.py (hunk 14 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py]: Decided to rename, add comment, make an issue

File: src/porepy/time_stepper/time_stepper.py (hunk 17 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_stepper.py]: made unconditional

File: src/porepy/time_stepper/time_step_constraint.py [hunk 31]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class CourantTimeStepConstraint(TimeStepConstraint):
+    """Adjust dt to follow the provided target CFL = max(v) * dt / min(x), where v is
+    the fluid velocity and x is the cell diameter.
+
+    Expects "model" (:class:`pp.PorePyModel`) in context.
+
+    Developer note: This is experimental and was not tested in practical simulations.
+    Use with causion.
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py]:
```suggestion
    Use with caution.
```

File: src/porepy/time_stepper/time_step_constraint.py [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+        """Velocity tolerance, below treated as zero."""
+
+    def suggest_dt(self, dt: float, context: dict) -> float:
+        model = context.get("model", None)
+        if not isinstance(model, pp.SolutionStrategy):
+            raise ValueError("CourantTimeStepConstraint requires 'model' in context.")
+
+        if len(model.fluid.phases) != 1 or len(model.fluid.components) != 1:
+            raise NotImplementedError(
+                "CourantTimeStepConstraint is not thought-through for multicomponent "
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py]:
```suggestion
                "CourantTimeStepConstraint is not thought through for multicomponent "
```

File: src/porepy/time_stepper/time_step_constraint.py [hunk 33]
```diff
@@ -0,0 +1,182 @@
...
+            )
+
+        dt = float("inf")
+        for subdomain in model.mdg.subdomains():
+            domains = [subdomain]
+            v_op = model.darcy_flux(domains)
+            v = np.max(abs(model.equation_system.evaluate(v_op)))
+            # As darcy_flux docstring suggests, its unit is [m^2 / s], because it is not
+            # scaled by the mobility term. Dividing it by the fluid viscosity ensures
+            # the right unit for a single-phase problem.
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_constraint.py:175]: This assumes constant viscosity, right? Add to docstring?

File: src/porepy/time_stepper/time_step_control.py [hunk 34]
```diff
@@ -1,400 +1,196 @@
...
-                sim_times,
-                rtol=self.rtol,
-                atol=self.atol,
-            )
-            if not is_compatible:
-                msg = (
-                    "Mismatch between the time step and scheduled time. Make sure the"
-                    " two are compatible, or consider adjusting the tolerances of the"
-                    " sanity check."
+            if dt_init is None:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_control.py:128]: Should we do elif schedule is arraylike - else raise ValueError?

File: src/porepy/time_stepper/time_step_control.py [hunk 35]
```diff
@@ -1,400 +1,196 @@
...
+        self.atol = atol
+        """Snapping time. If the time difference is below it, treats two time points
+        as equal.
 
-            # Sanity check for maximum number of recomputation attempts
-            if recomp_max <= 0:
-                raise ValueError("Number of recomputation attempts must be > 0.")
+        """
 
+        self.advanced_schedule: Schedule
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_control.py]: Why is the attribute called advanced_schedule, not simply schedule?

File: src/porepy/time_stepper/time_step_control.py [hunk 36]
```diff
@@ -413,312 +209,95 @@ def __init__(
...
-                step. A warning is raised if `iterations` is given when
-                `recompute_solution = True` or `constant_dt = True`.
-            recompute_solution: Whether the solution needs to be recomputed or not. If
-                True, then the time step is multiplied by `recomp_factor`. If False,
-                then the time step will be tuned accordingly.
-
-        Returns:
-            Next time step if time < final_time. None, otherwise.
+    @property
+    def schedule(self) -> np.ndarray:
```
Comment (IvarStefansson) [inline src/porepy/time_stepper/time_step_control.py]: Since this is not a Schedule, I'm wondering if schedule may not be the best name. Is that related to backwards compatibility?

**Review by IvarStefansson — APPROVED** [review]
Great work on this one, @Yuriyzabegaev! Approving now, feel free to merge if you are confident about resolving comments. Please let me know if you wish to discuss anything.

File: src/porepy/time_stepper/time_step_constraint.py (hunk 33 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_constraint.py:175]: added, thanks.

File: src/porepy/time_stepper/time_step_control.py (hunk 34 above)
Comment (Yuriyzabegaev) [reply src/porepy/time_stepper/time_step_control.py:128]: added, thanks

**Comment by Yuriyzabegaev** [comment]
the final resolution was to introduce a breaking change: `time_manager.schedule` is an object and not an array anymore. Use `time_manager.schedule.get_array()` to preserve old behavior. This might break runscripts, but we decided that it is better than introducing confusion by `time_manager.advanced_schedule` (which is purged).

This is reflected in the changelog.
